pkg/pillar/cmd/zedagent/handlenetworkinstance.go

```diff
@@ -535,42 +539,97 @@ func protoEncodeAppFlowMonitorProto(ipflow types.IPFlow) *flowlog.FlowMessage {
 	return pflows
 }
 
-func sendFlowProtobuf(protoflows *flowlog.FlowMessage) {
-
-	flowQ.PushBack(protoflows)
-
-	for flowQ.Len() > 0 {
-		ent := flowQ.Front()
-		pflowsPtr := ent.Value.(*flowlog.FlowMessage)
-
-		data, err := proto.Marshal(pflowsPtr)
-		if err != nil {
-			log.Errorf("FlowStats: SendFlowProtobuf proto marshaling error %v", err) // XXX change to fatal
+func flowlogTask(ctx *zedagentContext, flowlogQueue <-chan *flowlog.FlowMessage) {
+	wdName := agentName + "flowlog"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	var (
+		iteration    int
+		retryMsg     *flowlog.FlowMessage
+		retryAttempt int
+		retryTimer   *time.Timer
+	)
+	publish := func(msg *flowlog.FlowMessage) {
+		start := time.Now()
+		log.Function("flowlogTask got message")
+		var retry bool
+		err := publishFlowMessage(msg, iteration)
+		if err == nil {
+			iteration++
+		} else {
+			log.Error(err)
+			if (100*len(flowlogQueue))/cap(flowlogQueue) > 90 {
+				// More than 90% of the queue is used, start dropping instead of retrying.
+				log.Warnf("flowlogTask: dropped flow message: %+v", msg.Scope)
```

Comment:
Would make sense to add some metrics struct and include that in the metrics proto  message so that one can see the number of messages successfully send and number of messages dropped.